claat/parser/md/parse_test.go

```diff
@@ -0,0 +1,339 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package md
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/googlecodelabs/tools/claat/types"
+
+	"golang.org/x/net/html"
+)
+
+const empty = ""
+
+var errNonNil = errors.New(empty)
+
+func buildParserWithStep(markup string) *parserState {
+
+	ps := parserState{
+		tzr: html.NewTokenizer(strings.NewReader(markup)),
+		c:   &types.Codelab{},
+	}
+	ps.currentStep = ps.c.NewStep("title")
+	return &ps
+}
+
+func TestHandleCodelabTitle(t *testing.T) {
+	// Set up.
+	title := "Egret"
+	ps := buildParserWithStep(fmt.Sprintf("<h1>%s</h1>", title))
+	ps.advance()
+	handleCodelabTitle(ps)
+
+	if ps.c.Title != title {
+		t.Errorf("[%q] expected %v, got %v", title, title, ps.c.Title)
+	}
+}
+
+func TestProcessDuration(t *testing.T) {
+	type TestCase struct {
+		in  string
+		out time.Duration
+		err error
+	}
+	tests := []TestCase{
+		// Test an easy case.
+		TestCase{
+			in:  "1:00",
+			out: time.Duration(1) * time.Hour,
+			err: nil,
+		},
+		// Test an easy case with period delimiter.
+		TestCase{
+			in:  "1.00",
+			out: time.Duration(1) * time.Hour,
+			err: nil,
+		},
+		// Test a weird number case.
+		TestCase{
+			in:  "13:37",
+			out: time.Duration(817) * time.Minute,
+			err: nil,
+		},
+		// Test that minutes can be longer than an hour.
+		TestCase{
+			in:  "0:90",
+			out: time.Duration(90) * time.Minute,
+			err: nil,
+		},
+		// Test zero.
+		TestCase{
+			in:  "0",
+			out: time.Duration(0),
+			err: nil,
+		},
+		// Test more than two digits in the hours place.
+		TestCase{
+			in:  "100:00",
+			out: time.Duration(100) * time.Hour,
+			err: nil,
+		},
+		// Test that hours must be present.
+		TestCase{
+			in:  ":25",
+			out: 0,
+			err: errNonNil,
+		},
+		// Test that two minutes digits are required.
+		TestCase{
+			in:  "2:2",
+			out: 0,
+			err: errNonNil,
+		},
+		// Test that two minutes digits are required.
+		TestCase{
+			in:  "4:444",
+			out: 0,
+			err: errNonNil,
+		},
+		// Test that zero is the only simple scalar allowed.
+		TestCase{
+			in:  "600613",
+			out: 0,
+			err: errNonNil,
+		},
+		// Test an empty string.
+		TestCase{
+			in:  "",
+			out: 0,
+			err: errNonNil,
+		},
+		// Test complete nonsense.
+		TestCase{
+			in:  "This isn't even remotely close to resembling a correct duration string.",
+			out: 0,
+			err: errNonNil,
+		},
+	}
+
+	for _, tc := range tests {
+		out, err := processDuration(tc.in)
+		if tc.err == errNonNil {
+			// Expect an error.
+			if err == nil {
+				t.Errorf("[%q] expected non-nil error", tc.in)
+			}
+		} else {
+			if err != nil {
+				t.Errorf("[%q] expected nil error, got %v", tc.in, err)
+			} else if out != tc.out {
+				t.Errorf("[%q] expected %v, got %v", tc.in, tc.out, out)
+			}
+		}
```

Comment:
a switch would be more readable here. also, please don't use `errNonNil`, it doesn't give you any advantage, just makes things less readable:

```go
switch {
case tc.err != nil && err == nil:
  t.Errorf("%d: processDuration(%q) = %v; want error", i, tc.in, out)
case tc.err == nil && err != nil:
  t.Errorf("%d: processDuration(%q) = %v; want %v", i, tc.in, err, tc.out)
case out != tc.out:
  t.Errorf("%d: processDuration(%q) = %v; want %v", i, tc.in, out, tc.out)
}